src/EventStore.Core/DataStructures/ProbabilisticFilter/PersistentBloomFilters/BloomFilterAccessor.cs

```diff
@@ -0,0 +1,226 @@
+			Ensure.Positive(hashSize, nameof(hashSize));
+			Ensure.Positive(pageSize, nameof(pageSize));
+
+			if (logicalFilterSize < MinSizeKB * 1000 || logicalFilterSize > MaxSizeKB * 1000) {
+				throw new ArgumentOutOfRangeException(
+					nameof(logicalFilterSize),
+					$"size should be between {MinSizeKB:N0} KB and {MaxSizeKB:N0} KB inclusive");
+			}
+
+			if (Header.Size > cacheLineSize)
+				throw new Exception("header needs to fit in a cache line");
+
+			if (pageSize % cacheLineSize != 0)
+				throw new ArgumentException(
+					$"PageSize {pageSize} must be a multiple of CacheLineSize {cacheLineSize}");
+
+			LogicalFilterSize = logicalFilterSize;
+			LogicalFilterSizeBits = logicalFilterSize * 8;
+			CacheLineSize = cacheLineSize;
+			HashSize = hashSize;
+			PageSize = pageSize;
+			_onPageDirty = onPageDirty;
+			_log = log;
+
+			// add room for filter bits, rounded down to nearest cacheline
+			FileSize = GetBytePositionInFile(LogicalFilterSizeBits).RoundDownToMultipleOf(cacheLineSize);
+			FileSize += cacheLineSize;
+
+			// header is not part of the pages
+			// last page may be a partial page
+			var dataSize = FileSize - CacheLineSize;
+			NumPages = dataSize / PageSize;
+			if (dataSize % PageSize != 0)
+				NumPages++;
+		}
+
+		public long LogicalFilterSize { get; }
+		public long LogicalFilterSizeBits { get; }
+		public int CacheLineSize { get; }
+		public int HashSize { get; }
+		public int PageSize { get; }
+		public long FileSize { get; }
+
+		// last page may be a partial page
+		public long NumPages { get; }
+
+		// FileSize is definitely a multiple of CacheLineSize
+		public long NumCacheLines => FileSize / CacheLineSize;
+
+		// Pointer to the beginning of the file
+		private byte* _pointer;
+		public byte* Pointer {
+			get => _pointer;
+			set {
+				if ((long)value % CacheLineSize != 0) {
+					//qq consider whether to keep this.. throw? log error?
+					// do at least test out how the error message comes out
+					_log.Error("Pointer {pointer} was not aligned to a cacheline {cacheLineSize}", (long)value, CacheLineSize);
+					throw new InvalidOperationException($"Pointer {(long)value} was not aligned to a cacheline {CacheLineSize}");
+				}
+				_pointer = value;
+			}
+		}
+
+		// converts a bit position in the logical filter into a byte position in the file
+		public long GetBytePositionInFile(long bitPosition) {
+			var bytePositionInFile = bitPosition / 8;
+
+			// account for 4 bytes at the end of each cache line being a hash
+			bytePositionInFile += bytePositionInFile / (CacheLineSize - HashSize) * HashSize;
+
+			// first cache line is reserved for the header
+			bytePositionInFile += CacheLineSize;
+
+			return bytePositionInFile;
+		}
+
+		// converts page number into byte position of the page in the file
+		public (long PositionInFile, int PageSize) GetPagePositionInFile(int pageNumber) {
+			// first cache line is reserved for the header
+			var positionInFile = CacheLineSize + (pageNumber * PageSize);
```

Comment:
@timothycoleman it looks like the multiplication could overflow an int for a 4GB bloom filter

Reply:
🤦‍♂️ thats a jolly good spot